op-e2e/actions: Add Holocene FP action tests

geoknee opened the pull request:
This PR adds action tests to cover consensus changes introduced by the Holocene hardfork. The action tests also run against the fault proof program.

A few changes to consensus code are introduced as well, which were identified during debugging of the action tests:
- Handle `rpc.Error`s directly instead of relying on `eth.InputError`s (https://github.com/ethereum-optimism/optimism/pull/12520/commits/2f6922dc678765ec6219ef3a5c32034b0e6e45d2 https://github.com/ethereum-optimism/optimism/pull/12520/commits/e6f8a0904ce4cdcf3bb777b89957aa68e21ca573)
The fault proof program's L2 Engine API doesn't return `eth.InputError`s,
like the sources engine client, but directly returns `rpc.Error`s.
So instead of relying on this translation, derivers need to deal
directly with `rpc.Error`s.
- [op-program: Don't close deriver when derivation goes idle](https://github.com/ethereum-optimism/optimism/pull/12520/commits/89c59724fbe7ee51fb16fb2cf7ac03503d609252)
The FCU event will close the deriver in time. If we cannot reach the
target otherwise, it doesn't matter that we don't close.
In Holocene, if the last attributes are invalid, deposit-only attributes
are retried. Due to event race conditions, the deriver may then already
get idle while the engine is still processing events. So the program can
close without having updated the safe head to pending/local yet.

*- @sebastianst*

Closes #12449

geoknee committed:
Add Holocene action tests

geoknee commented on op-e2e/actions/proofs/holocene_test.go:
```diff
@@ -0,0 +1,193 @@
+	{name: "case-3", blocks: []uint{2, 2, 1, 3}, frames: []uint{0, 1, 2, 3}, safeHeadPreHolocene: 3, safeHeadHolocene: 0}, // duplicate block
+	{name: "case-4", blocks: []uint{1, 2, 3}, frames: []uint{2, 1, 0}, safeHeadPreHolocene: 3, safeHeadHolocene: 0},       // bad frame ordering
+	{name: "case-5", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 0, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 0},    // duplicate frames
+	{name: "case-6", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 0, safeHeadHolocene: 0,
+		isSpanBatch: true, blockModifiers: []actionsHelpers.BlockModifier{nil, blockFudger, nil}}, // partially invalid span batch (invalid payload)
+	{name: "case-7", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 1, safeHeadHolocene: 1,
+		isSpanBatch: false, blockModifiers: []actionsHelpers.BlockModifier{nil, blockSpudger, nil}}, // partially invalid singular batch channel (invalid batch)
+	{name: "case-8", blocks: []uint{1, 2, 3}, frames: []uint{0, 1, 2}, safeHeadPreHolocene: 3, safeHeadHolocene: 3,
+		isSpanBatch: true, blockModifiers: []actionsHelpers.BlockModifier{nil, blockSpudger, nil}}, // partially invalid span batch (invalid batch?)
```
So far these cases are failing to exercise any different behaviour under Holocene.

sebastianst replied:
The invalid payload handling isn't implemented yet. This doesn't happen at the batch stage, but later in the engine queue. Do we have tests that invalidate the _batch_ not the payload?

geoknee replied:
As discussed on our call today, I"m going to look into adding such tests next.

geoknee replied:
These have now been added.

geoknee commented on op-program/client/l2/engineapi/l2_engine_api.go:
```diff
@@ -126,11 +129,12 @@ func (ea *L2EngineAPI) IncludeTx(tx *types.Transaction, from common.Address) err
 	}
-	if ea.l2ForceEmpty {
-		ea.log.Info("Skipping including a transaction because e.L2ForceEmpty is true")
-		// t.InvalidAction("cannot include any sequencer txs")
```
This commented out line looks out of place, it is an action test helper.

sebastianst commented:
I've based this branch on top of https://github.com/ethereum-optimism/optimism/pull/12724 which fixes all Holocene behavior to a prod ready version. So we can adapt tests to do proper assertions, e.g. using safe instead of pending safe.

sebastianst committed:
fix invalid batch tests

sebastianst committed:
Handle rpc.Errors directly instead of relying on eth.InputErrors

The fault proof program's L2 Engine API doesn't return eth.InputErrors,
like the sources engine client, but directly returns rpc.Errors.
So instead of relying on this translation, derivers need to deal
directly with rpc.Errors.

sebastianst committed:
In TryBackupUnsafeReorg, only reset on InvalidForkchoiceState error code

sebastianst committed:
Add logs

clabby reviewed (approved):
The existing tests here LGTM, and kona is passing all of them as well 🥳 

Some other tests that would be great to follow up on:
1. Fault prove the holocene activation block, as well as holocene activation block + 1, for added coverage over the initial behavior of the payload building code.
2. Coverage over the new channel size restriction in the channel stage, where compressed bytes must be < the max rlp bytes per channel protocol constant.
3. @sebastianst's described edge case where we did need the frame ordering restriction in the actual `Channel` datatype, and not just the frame queue.

ajsutton commented on op-program/client/driver/program.go:
```diff
@@ -63,9 +63,9 @@ func (d *ProgramDeriver) OnEvent(ev event.Event) bool {
-		d.closing = true
-		d.logger.Info("Derivation complete: no further data to process")
+		// We dont't close the deriver yet, as the engine may still be processing events to reach
+		// the target. A ForkchoiceUpdateEvent will close the deriver when the target is reached.
```
We aren't properly handling the case where there isn't sufficient data to reach the target.  ie we've reached L1 head and don't have batch data to advance and the sequencing window hasn't completed.  The comment here is fairly misleading as there is nothing to guarantee that a `ForkchoiceUpdateEvent` will ever happen that reaches the target.

Previously we'd get:

```
INFO [11-04|06:24:20.630] Derivation complete: no further data to process
INFO [11-04|06:24:20.631] Validating claim                         head=9e5418..f77ece:123833816 output=7105fc..805f0f claim=97cee0..d4bbd5
CRIT [11-04|06:24:20.631] Application failed                       err="invalid claim: claim: 0x97cee0334b05467175319ce900ae78b1db6c9949f6b3b62f399eb2b40cd4bbd5 actual: 0x7105fcfdba9c303605cf3e7dec8eb5abf5ad7438f207f61e74ce3fc5e2805f0f"
```

whereas now we get:
```
INFO [11-04|06:23:20.007] BatchMux: activating pre-Holocene stage during reset origin=3625ea..26256e:20495701
INFO [11-04|06:23:20.016] Derivation complete: no further L1 data to process
CRIT [11-04|06:23:20.016] Application failed                       err="failed to run program to completion: exhausted events before completing program"
```

It happens that we still return an exit code of 1 so it "works" for invalid outputs. However the honest actor is expected to post the latest safe head if the proposed output root is from a L2 block after that and so the safe head needs to be considered valid even if the actual L2 block number hasn't been reached yet, but this change makes op-program return invalid in that case.

This case gets picked up by the cannon e2e tests which don't run pre-merge but the failures are https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/70897/workflows/43c44f11-8eec-4bec-8746-64f1f464bd99/jobs/2921089/tests
You can run the pipeline with the param `fault_proofs_dispatch` to get CI to run the tests prior to merge.

ajsutton replied:
It does seem odd to me that we're getting a DerivationIdle here. In op-node that will trigger it to try and update the finalized block because it assumes that all L2 blocks derived from the L1 block have now been imported.  That no longer appears to be the case. That will all come out in the wash as it will just update the L2 finalized later anyway (if it makes a difference in the first place), but it may be worth reviewing if we really should be firing that idle event - seems like it's meaning has changed now.

ajsutton replied:
I think https://github.com/ethereum-optimism/optimism/pull/12804 would be the right way to fix this if we do keep this change to the meaning of the `DerivationIdle` event.

protolambda replied:
The exhaust-data error case results in a generic error, and then the programs exists with code 2, since it doesn't wrap `ErrClaimNotValid`. I suggest we make the exhaustion case wrap around `ErrClaimNotValid` to make the program output consistent with what it used to do.

protolambda replied:
I think #12804, now that I have discussed with Seb, is also a valid solution to return the correct exit code. The claim-check, since it constructs the expected-claim based on the safe-head, is a valid comparison against the claimed safe-head. Exhaustion doesn't need to be a separately returned claim error.

sebastianst replied:
And if the program run does reach the target, it will always exit early here:
```go
	case engine.ForkchoiceUpdateEvent:
		if x.SafeL2Head.Number >= d.targetBlockNum {
			d.logger.Info("Derivation complete: reached L2 block", "head", x.SafeL2Head)
			d.closing = true
		}
```
So the valid claim check is guaranteed to fail if we exhaust all events, while the last event _wasn't_ a `engine.ForkchoiceUpdateEvent`.

sebastianst replied:
Merged #12803 and #12804 and rebased this PR on top of it.

ajsutton reviewed (dismissed):
This breaks the fault proof system unfortunately.  I'll have a play and see if I can come up with a fix.

protolambda commented on op-service/sources/engine_client.go:
```diff
@@ -100,28 +95,15 @@ func (s *EngineAPIClient) ForkchoiceUpdate(ctx context.Context, fc *eth.Forkchoi
-		if rpcErr, ok := err.(rpc.Error); ok {
-			code := eth.ErrorCode(rpcErr.ErrorCode())
-			switch code {
-			case eth.InvalidParams, eth.InvalidForkchoiceState, eth.InvalidPayloadAttributes:
```
I believe that `InvalidParams` (generic RPC error), will result in a temporary-error loop now, but that's fine, since it implies the op-node/op-program speaks a different RPC than the engine, and stalling/aborting is fine in that case. It wouldn't happen during a correctly set up CL / EL combination.

sebastianst replied:
yep, it's just the operational difference between a retry loop vs an instant crash

protolambda commented on op-program/client/driver/program_test.go:
```diff
@@ -104,14 +104,6 @@ func TestProgramDeriver(t *testing.T) {
 			require.NoError(t, p.result)
 		})
 	})
-	// on exhaustion of input data: stop without error
-	t.Run("deriver idle", func(t *testing.T) {
```
Adrian is adding a test for this in 12803, nice

protolambda commented on op-e2e/actions/proofs/holocene_frame_test.go:
```diff
@@ -0,0 +1,153 @@
+
+		t.Log("Safe head progressed as expected", "l2SafeHeadNumber", l2SafeHead.Number)
+
+		if safeHeadNumber := l2SafeHead.Number; safeHeadNumber > 0 {
+			env.RunFaultProofProgram(t, safeHeadNumber, testCfg.CheckResult, testCfg.InputParams...)
+		}
```
I think this shouldn't be in an if statement, and run unconditionally

protolambda commented on op-e2e/actions/proofs/holocene_batches_test.go:
```diff
@@ -0,0 +1,144 @@
+			env.Miner.ActL1IncludeTxByHash(env.Batcher.LastSubmitted.Hash())(t)
+			env.Miner.ActL1EndBlock(t)
+
+			// Finalize the block with the first channel frame on L1.
+			env.Miner.ActL1SafeNext(t)
+			env.Miner.ActL1FinalizeNext(t)
```
You can safely remove these few lines

protolambda commented on op-e2e/actions/proofs/holocene_batches_test.go:
```diff
@@ -0,0 +1,144 @@
+
+		t.Log("Safe head progressed as expected", "l2SafeHeadNumber", l2SafeHead.Number)
+
+		if safeHeadNumber := l2SafeHead.Number; safeHeadNumber > 0 {
+			env.RunFaultProofProgram(t, safeHeadNumber, testCfg.CheckResult, testCfg.InputParams...)
+		}
```
Unconditional also?

Maybe don't assume the output 0 or non-zero, but use the testcase input to determine whether or not to run the FP?

protolambda reviewed (commented):
LGTM for most changes. Error handling of RPC is good, but need the PR from Adrian for exit-code behavior to be accurate. And some test cases can be refined, I think the FP should always run. And asserting a block-number after the L1-sync would be nice.

sebastianst committed:
include genesis FPP tests
